src/cargo/ops/lockfile.rs

```diff
@@ -81,6 +81,12 @@ pub fn write_pkg_lockfile(ws: &Workspace, resolve: &Resolve) -> CargoResult<()>
         if has_crlf_line_endings(&orig) {
             out = out.replace("\n", "\r\n");
         }
+        // Forward compatibility: don't change lockfile if `orig` uses
+        // rootless format from the future.
+        if orig.starts_with("[[package]]") {
+            assert!(out.starts_with("[root]"));
+            out = format!("[[package]]{}", &out["[root]".len()..]);
+        }
```

Comment:
This is interesting. 

This is the case where we've read a Cargo.toml without `[root]` from the future. Presumably we want to preserve this rootless format. 

But this approach may change the ordering of packages in the lockfile! This is not the best behavior. Hm, I think I can actually solve this...